Fix get actor timeout multiplier

VarunBhandary opened the pull request:
# Fix timeout multiplier bug in Ray state API

## Why are these changes needed?

This PR fixes a critical bug in the Ray state API where small user-specified timeouts (like 1 second) would result in 0-second timeouts after applying the server timeout multiplier, causing assertion errors.

**Problem:**
- When users set `timeout=1` second in `GetApiOptions` or `ListApiOptions`, the server timeout multiplier (0.8) would reduce it to `int(1 * 0.8) = 0`
- This triggered the assertion error: `"0 second timeout is not supported."`
- Users couldn't use the Ray state API with short timeouts

**Root Cause:**
The timeout calculation used `int(self.timeout * self.server_timeout_multiplier)` which truncates to 0 for small timeouts, making the API unusable for quick operations.

**Solution:**
Changed the timeout calculation to `max(1, int(self.timeout * self.server_timeout_multiplier))` in both:
- `ListApiOptions.__post_init__()`
- `GetApiOptions.__post_init__()`

This ensures the effective timeout is always at least 1 second while preserving the server timeout multiplier behavior for larger timeouts.

**Impact:**
- ✅ Fixes crash when users specify timeout=1 second
- ✅ Maintains backward compatibility for normal timeout values
- ✅ Preserves the server timeout multiplier logic for preventing HTTP timeouts
- ✅ Both `GetApiOptions` and `ListApiOptions` now behave consistently

## Related issue number

Fixes timeout multiplier edge case in Ray state API.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] No doc changes needed - this is a bug fix maintaining existing API
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests - Created comprehensive tests for edge cases
   - [x] Integration tests - Verified fix works with Ray state API
   - [x] Manual testing - Tested both timeout=1 and normal timeout values
   - [x] Regression testing - Ensured existing functionality still works

**Test Results:**
- ✅ `GetApiOptions(timeout=1)` now works correctly (returns 1s instead of crashing)
- ✅ `ListApiOptions(timeout=1)` now works correctly (returns 1s instead of crashing)  
- ✅ Both options maintain consistency with the same timeout values
- ✅ Normal timeout values (>1s) continue to work as expected
- ✅ Server timeout multiplier still prevents HTTP timeouts as intended

vb-dbrks committed:
Fix timeout multiplier issue in GetApiOptions

- Add server_timeout_multiplier field to GetApiOptions
- Add __post_init__ method to apply timeout multiplier consistently
- Ensure GetApiOptions applies same 0.8 timeout multiplier as ListApiOptions
- Add test to verify the fix works correctly

This fixes the issue where get_actor() with timeout=1 would fail even
when the operation took less than 1 second, because GetApiOptions didn't
apply the same server timeout multiplier that ListApiOptions uses.

Fixes #54153

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

vb-dbrks committed:
Fix formatting in test file

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

vb-dbrks committed:
Fix timeout multiplier bug in Ray state API

- Fix zero timeout issue when user timeout * multiplier < 1
- Apply max(1, calculated_timeout) to prevent assertion errors
- Ensure GetApiOptions and ListApiOptions behavior consistency
- Maintain backward compatibility for normal timeout values

Fixes issue where timeout=1 second would result in 0 seconds
after applying the 0.8 server_timeout_multiplier, causing
'0 second timeout is not supported' assertion errors.

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

vb-dbrks committed:
Fix GetApiOptions timeout multiplier and build script issues

- Apply server_timeout_multiplier to GetApiOptions similar to ListApiOptions
- Fix has_conflicting_filters to handle None filters properly
- Remove redundant import in test_get_actor_timeout_multiplier test
- Add proper type annotations and static method decorators
- Refactor JobState to handle None JobDetails gracefully

This ensures consistent timeout behavior between GetApiOptions and
ListApiOptions, and resolves build script failures.

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

vb-dbrks committed:
Apply Black formatting fixes

- Add blank lines before class definitions
- Reformat dictionary comprehension for better readability

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

dayshah reviewed (commented):
This a great high quality contribution, thanks a lot for the beautiful PR description!

i think there's some additional pydantic related changes you're making? Could you split those out into a separate PR?

VarunBhandary committed:
Update python/ray/tests/test_state_api.py

Co-authored-by: Dhyey Shah <dhyey2019@gmail.com>
Signed-off-by: DataErrata <varun.bhandary@gmail.com>

VarunBhandary committed:
Update python/ray/tests/test_state_api.py

Co-authored-by: Dhyey Shah <dhyey2019@gmail.com>
Signed-off-by: DataErrata <varun.bhandary@gmail.com>

vb-dbrks committed:
Fix GetApiOptions timeout multiplier inconsistency

Apply server_timeout_multiplier consistently between GetApiOptions and
ListApiOptions to prevent timeout failures. GetApiOptions now uses the
same 0.8 multiplier and max(1, ...) logic as ListApiOptions.

Changes:
- Add server_timeout_multiplier field to GetApiOptions
- Apply max(1, int(timeout * multiplier)) to prevent 0-second timeout
- Remove unrelated pydantic changes per reviewer feedback
- Add test case verifying consistent timeout behavior
- Remove redundant ray.init() call from test

This fixes get_actor() calls failing with timeout=1 even when the
operation completes quickly, ensuring consistent timeout behavior
across the state API.

Fixes: https://github.com/ray-project/ray/issues/54153
Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

VarunBhandary commented on python/ray/util/state/common.py:
```diff
@@ -105,17 +107,20 @@ def memory(x: int):
                 event["created_ms"] = Humanify.timestamp(event["created_ms"])
         return events
 
+    @staticmethod
```
thank you for the feedback. reverted and might create a separate pr - but don't think the changes are strictly necessary .

VarunBhandary commented on python/ray/util/state/common.py:
```diff
@@ -526,7 +543,7 @@ class PlacementGroupState(StateSchema):
     #: - RESCHEDULING: The placement group is rescheduling because some of
     #:   bundles are dead because they were on dead nodes.
-    state: TypePlacementGroupStatus = state_column(filterable=True)
+    state: "TypePlacementGroupStatus" = state_column(filterable=True)
```
thank you for the feedback. reverted.

VarunBhandary commented on python/ray/util/state/common.py:
```diff
@@ -172,7 +177,7 @@ def has_conflicting_filters(self) -> bool:
         # and ther other with '!=' predicate
         equal_filters = {}
-        for filter in self.filters:
+        for filter in self.filters or []:
```
thank you for the feedback. Reverted as suggested.

VarunBhandary committed:
Merge branch 'master' into fix-get-actor-timeout-multiplier

vb-dbrks committed:
fix(test): Remove whitespace from blank line in test_state_api.py

Fix ruff linting error W293 by removing trailing whitespace from blank line
in the test_get_actor_timeout_multiplier test docstring.

This addresses a minor linting issue unrelated to the core timeout
multiplier fix but ensures CI passes cleanly.

Signed-off-by: Varun Bhandary <varun.bhandary@databricks.com>

VarunBhandary committed:
Merge branch 'master' into fix-get-actor-timeout-multiplier

dayshah reviewed (approved):
nice, this lgtm!
